src/sensors/trilateration.cpp

```diff
@@ -0,0 +1,368 @@
+            real_z_relative_to_pinger);
+    const double pos_two_depth_difference = fabs(position_two_rectified[2] -
+            real_z_relative_to_pinger);
+    if (distances[0] > 0 && position_one_rectified[2] <
+            -1 * pinger_depth && pos_one_depth_difference < 2)
+    {
+        position_readings.push_back(position_one_rectified);
+        distance_readings.push_back(distances[0]);
+    }
+    else
+    {
+        ROS_INFO_STREAM("Position one invalid:\n" << position_one_rectified <<
+                "\nDistance: " << distances[0]);
+    }
+
+    if (distances[1] > 0 && position_two_rectified[2] <
+            -1 * pinger_depth && pos_two_depth_difference < 2)
+    {
+        position_readings.push_back(position_two_rectified);
+        distance_readings.push_back(distances[1]);
+    }
+    else
+    {
+        ROS_INFO_STREAM("Position two invalid:\n" << position_two_rectified <<
+                "\nDistance: " << distances[1]);
+    }
+
+    /*
+     * Order them based on how close the depth reading is if there are two
+     * valid readings still. If the order swaps, swap the distance measurements
+     * as well so that they are properly displayed in the message.
+     */
+    if (position_readings.size() > 1 && pos_two_depth_difference <
+            pos_one_depth_difference)
+    {
+        position_readings.push_back(position_readings[0]);
+        position_readings.erase(position_readings.begin());
+
+        distance_readings.push_back(distance_readings[0]);
+        distance_readings.erase(distance_readings.begin());
+    }
+
+    /*
+     * Publish all of the rectified measurements.
+     */
+    robosub::PositionsStamped all_positions_msg;
+    geometry_msgs::Vector3 position;
+    all_positions_msg.header.stamp = ros::Time::now();
+
+    all_positions_msg.distances.push_back(distances[0]);
+    all_positions_msg.distances.push_back(distances[1]);
+
+    position.x = position_one[0];
+    position.y = position_one[1];
+    position.z = position_one[2];
+    all_positions_msg.positions.push_back(position);
+    position.x = position_two[0];
+    position.y = position_two[1];
+    position.z = position_two[2];
+    all_positions_msg.positions.push_back(position);
+
+    position_all_pub.publish(all_positions_msg);
+
+    /*
+     * The current position readings are all valid and ordered by confidence.
+     * Add them to a position message.
+     */
+    robosub::PositionsStamped position_msg;
+    position_msg.header.stamp = ros::Time::now();
+    for (unsigned int i = 0; i < position_readings.size(); ++i)
+    {
+        position_msg.distances.push_back(distance_readings[i]);
+
+        position.x = position_readings[i][0];
+        position.y = position_readings[i][1];
+        position.z = position_readings[i][2];
+        position_msg.positions.push_back(position);
+    }
+
+    if (position_readings.size())
+        position_pub.publish(position_msg);
```

Comment:
Depending upon who is listening to this topic, would it be reasonable for them to want an empty message given no valid options rather than no message?